### PR TITLE
Another try at account balancing

### PR DESCRIPTION
**Pull request by bittelc**
## Problem

[Asana Link](https://app.asana.com/0/998359196895599/1123355154449933)

Transactions that are processed by validators fail to update the underlying receiver and sender accounts. This means that no sender is ever credited and no receiver is ever credited, if there are Validators registered with the Supervisor.

## Solution

Prior to sharding to the Validators for processing, transactions are pre-processed by the Supervisor to update each of the sender and receiver accounts to reflect the balance of the account with the transaction applied to the accounts'.

## Testing Done and Results

<details>
<summary>crediting from sender</summary>

```
╰─➤  curl http://127.0.0.1:80/account/HDzLGL98C4vKtVWb3qzm92C2LX2V5kNhXR
{"nonce":61,"address":"HDzLGL98C4vKtVWb3qzm92C2LX2V5kNhXR","balance":4800,"storageRoot":"B64AD48329AF45557DF97F82B4BFEB9411AD80C2F1C6DBAE0EE4D12AC55B4575","publicKey":"8YHKuCEDvZ+MEyEyQOA/4NAk6Q8SeB/vpdqv30mqODMaQQaMFaE=","EBalances":null}
╭─bittelc@coles-mbp ~/go/src/github.com/herdius/herdius-blockchain-api  ‹master*›
╰─➤  go run client.go
2019/05/21 16:19:38 endpoint: http://localhost:80/tx
2019/05/21 16:19:38 Sender Address: HDzLGL98C4vKtVWb3qzm92C2LX2V5kNhXR
2019/05/21 16:19:38 Receiver Address: HPNMnZc9eNA7PzEMRWVqXwzPqieSRLzuyf
2019/05/21 16:19:39 HTxdc95d8548b94952bccaafbe37c251fd8f6f6b7bf
2019/05/21 16:19:39 success
╭─bittelc@coles-mbp ~/go/src/github.com/herdius/herdius-blockchain-api  ‹master*›
╰─➤  curl http://127.0.0.1:80/account/HDzLGL98C4vKtVWb3qzm92C2LX2V5kNhXR
{"nonce":62,"address":"HDzLGL98C4vKtVWb3qzm92C2LX2V5kNhXR","balance":4700,"storageRoot":"6AD32B4B711929F472DF397800875EFBFAB9BE256E26F172823C126C055A5E67","publicKey":"8YHKuCEDvZ+MEyEyQOA/4NAk6Q8SeB/vpdqv30mqODMaQQaMFaE=","EBalances":null}
```

</details>

### CODE REVIEW

File: storage/state/statedb/service.go [hunk 1]
```diff
@@ -68,6 +68,7 @@ func GetState(dir string) Trie {
 
 //NewTrie
 func NewTrie(hash common.Hash) (Trie, error) {
+	log.Println("hash:", hash)
```
Comment (bittelc) [inline storage/state/statedb/service.go]: This will be removed after debugging and peer review.

File: supervisor/service/service.go [hunk 2]
```diff
@@ -859,17 +859,307 @@ func (s *Supervisor) ShardToValidators(txs *txbyte.Txs, net *network.Network, st
 	}
 	numCbs = numGrps
 	fmt.Printf("Number of txs (%v), child blocks (%v), validators (%v)\n", numTxs, numCbs, numValds)
-
+	log.Println("stateRoot:", stateRoot)
```
Comment (bittelc) [inline supervisor/service/service.go]: Remove this

File: supervisor/service/service.go [hunk 3]
```diff
@@ -859,17 +859,307 @@ func (s *Supervisor) ShardToValidators(txs *txbyte.Txs, net *network.Network, st
...
+	if len(stateRoot) <= 0 {
+		return fmt.Errorf("Cannot process an empty stateRoot for the trie")
+	}
+	log.Println("length:", len(stateRoot))
+	log.Println("content:", stateRoot)
```
Comment (bittelc) [inline supervisor/service/service.go]: Remove this ^

File: supervisor/service/service_test.go [hunk 4]
```diff
@@ -307,12 +307,23 @@ func getTxSecp256k1Account(nonce int) transaction.Tx {
...
+}
+
+func TestShardToValidatorsTrue(t *testing.T) {
+	supsvc := &Supervisor{}
+	supsvc.AddValidator([]byte{1}, "add-01")
+	supsvc.AddValidator([]byte{1}, "add-02")
+	supsvc.SetWriteMutex()
+	txs := &txbyte.Txs{}
+	stateRoot := []byte{106, 211, 43, 75, 113, 25, 41, 244, 114, 223, 57, 120, 0, 135, 94, 251, 250, 185, 190, 37, 110, 38, 241, 114, 130, 60, 18, 108, 5, 90, 94, 103}
+	err := supsvc.ShardToValidators(txs, nil, stateRoot)
```
Comment (bittelc) [inline supervisor/service/service_test.go]: @lmahanand – I am looking for help getting this test to pass.
I am getting the following panic when I run this, and I'm having a hard time understanding why.

```
umber of txs (0), child blocks (2), validators (2)
2019/05/21 17:02:18 stateRoot: [106 211 43 75 113 25 41 244 114 223 57 120 0 135 94 251 250 185 190 37 110 38 241 114 130 60 18 108 5 90 94 103]
2019/05/21 17:02:18 length: 32
2019/05/21 17:02:18 content: [106 211 43 75 113 25 41 244 114 223 57 120 0 135 94 251 250 185 190 37 110 38 241 114 130 60 18 108 5 90 94 103]
2019/05/21 17:02:18 hash: [106 211 43 75 113 25 41 244 114 223 57 120 0 135 94 251 250 185 190 37 110 38 241 114 130 60 18 108 5 90 94 103]
--- FAIL: TestShardToValidatorsTrue (0.00s)
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x45cbcd8]

goroutine 6 [running]:
testing.tRunner.func1(0xc00011a200)
	/usr/local/go/src/testing/testing.go:830 +0x392
panic(0x4696f00, 0x4de9830)
	/usr/local/go/src/runtime/panic.go:522 +0x1b5
github.com/herdius/herdius-core/storage/state/statedb.NewTrie(0xf42919714b2bd36a, 0xfb5e87007839df72, 0x72f1266e25beb9fa, 0x675e5a056c123c82, 0xc00004ece0, 0x3, 0x3, 0x34)
	/Users/bittelc/go/src/github.com/herdius/herdius-core/storage/state/statedb/service.go:73 +0xe8
github.com/herdius/herdius-core/supervisor/service.(*Supervisor).ShardToValidators(0xc0000a7f28, 0xc0000a7f10, 0x0, 0xc0002c2000, 0x20, 0x20, 0x1, 0x2)
	/Users/bittelc/go/src/github.com/herdius/herdius-core/supervisor/service/service.go:868 +0x48b
github.com/herdius/herdius-core/supervisor/service.TestShardToValidatorsTrue(0xc00011a200)
	/Users/bittelc/go/src/github.com/herdius/herdius-core/supervisor/service/service_test.go:327 +0x2d3
testing.tRunner(0xc00011a200, 0x4883d68)
	/usr/local/go/src/testing/testing.go:865 +0xc0
created by testing.(*T).Run
	/usr/local/go/src/testing/testing.go:916 +0x35a
```

File: supervisor/service/service.go [hunk 5]
```diff
@@ -859,17 +859,304 @@ func (s *Supervisor) ShardToValidators(txs *txbyte.Txs, net *network.Network, st
...
+	if err != nil {
+		return fmt.Errorf("Error attempting to retrieve state db trie from stateRoot: %v", err)
+	}
 	previousBlockHash := make([]byte, 0)
 	txStr := transaction.Tx{}
 	txlist := &transaction.TxList{}
-	for _, txb := range *txs {
-		err := cdc.UnmarshalJSON(txb, &txStr)
+	tx := pluginproto.Tx{}
+	for i, txbz := range *txs {
```
Comment (lmahanand) [inline supervisor/service/service.go]: Can you please reuse this part from createSingularBlock() method. This for loop generates the updated **stateroot**. Put in a single method and reuse the method in both createSingularBlock() and ShardToValidators()

**Pull request merged**